### ISSUE TITLE
Enhancement: Support DDL statements (e.g. `DROP`)

### ISSUE THREAD

**Issue by adam-tokarski** [alanmcruickshank/sqlfluff#102]
Hi, I'm not sure whether DDL statements are supported, but I've got
```bash
== [db_mysql/docker-entrypoint-initdb.d/test.sql] FAIL
L:   1 | P:   1 | ???? | Found unparsable segment @L001P001: 'DROP VIEW  test_view...'
L:   1 | P:  22 | ???? | Found unparsable segment @L001P022: '\nDROP TABLE IF EXIST...'
```

for
```sql
DROP VIEW IF EXISTS test_view;
DROP TABLE IF EXISTS test_table;
```

Is there anything I can do with that?
Thanks in advance.

**Comment by alanmcruickshank** [comment]
Thanks for posting the issue @adam-tokarski .

There isn't any support for DDL statements yet, but I've got a similar use case where I'd like to implement some support for `GRANT` statements.

I'm happy to put together a first pass at some DDL statements. If you could pick 3/4 other DDL statements which others would you want?

**Comment by adam-tokarski** [comment]
All of them :wink: 
Frankly, you should start with the most popular ones. But until we will verify that from some statistics (from github for example), we can only guess which one are most common - I bet `DROP` and `CREATE` for tables and views. Maybe additionally `ALTER`? Also would be good to have covered `COMMIT` and `ROLLBACK`.

**Comment by adam-tokarski** [comment]
Alternatively it would be good to have an option to suppress statement,, to mark them as a ignored somehow.

**Comment by alanmcruickshank** [comment]
@barrywhart - do you fancy having a go at this one? I think `COMMIT` and `ROLLBACK` will be really easy, I'd be tempted to do `DROP`, `GRANT` and `REVOKE` after that, and then perhaps `CREATE` and `ALTER` last (as they might be the most complicated).

Don't worry *too much* about exactly how you structure where the definitions go in the dialect definition. I'd probably start with creating something like `DDLStatementSegment` which would then be one of the options in the `parse_grammar` of `StatementSegment`. Beyond that I'd encourage you to have a play with how the grammar is structured by looking at the objects which are already defined. For debugging, using the `parse` command line function is really helpful, especially using very high verbosities (using `-vvvvvvvvvvv`). Let me know how I else I can help.

**Comment by barrywhart** [comment]
Sure, I'll take a look.

**Comment by barrywhart** [comment]
Some things came up at work this week. I'll try and work on this next week. I have worked on parsers in the past, but always using a parser package like yacc or PLY, so I think I'll be okay after some "warm up". 🙂

**Comment by barrywhart** [comment]
Another pretty busy week, but this is near the top of my list. 🤞 I can start on it tomorrow (Friday) or next week.

**Comment by barrywhart** [comment]
@alanmcruickshank: I have a PR that adds `COMMIT`, `ROLLBACK`, and `DROP` support. For `GRANT` and `REVOKE`, I couldn't find an ANSI reference. I did find the [Postgres syntax](https://www.postgresql.org/docs/9.0/sql-grant.html), and it has a lot of variations. Do you have in mind supporting all these variations initially? It's doable, but looks like it could take a fair bit of time. I'll begin by adding support for the examples on that page _only_ (not the full syntax at the top).

**Comment by alanmcruickshank** [comment]
@barrywhart - thanks for the PR it looks great. I'm happy to support different variations progressively. Let's keep things in the ansi dialect so far as they don't cause conflicts, and let's add new variants of the syntax as they're needed. That way we can get features out faster.

Now that this is merged there are a few projects I want to battle test it against, but I think we're in a pretty good place.

@adam-tokarski - if you test against the current master branch is this enough for you to work with, or are there particular features still missing? If you could post the queries which fail here that would be awesome.

**Comment by adam-tokarski** [comment]
Awesome! Thanks for that.

There, I made little test and got 
```bash
sqlfluff lint test.sql
== [test.sql] FAIL
L:   3 | P:  35 |  PRS | Found unparsable segment @L003P035: '\n\n\nCREATE TABLE IF N...'
L:   7 | P:   5 | L003 | Indentation not consistent with line #5
L:   8 | P:   5 | L003 | Indentation not consistent with line #5
L:  10 | P:  41 |  PRS | Found unparsable segment @L010P041: '\n\nCREATE TABLE IF NO...'
L:  14 | P:   5 | L003 | Indentation not consistent with line #11
L:  15 | P:   5 | L003 | Indentation not consistent with line #11
L:  16 | P:   5 | L003 | Indentation not consistent with line #11
L:  17 | P:   5 | L003 | Indentation not consistent with line #11
L:  19 | P: 104 |  PRS | Found unparsable segment @L019P104: '\n\n/*\n *  Gathers all...'
L:  19 | P: 104 | L016 | Line is too long.
L:  26 | P:   5 | L003 | Indentation not consistent with line #20
L:  27 | P:   9 | L003 | Indentation not consistent with line #20
L:  28 | P:   5 | L003 | Indentation not consistent with line #20
L:  29 | P:   5 | L003 | Indentation not consistent with line #20
```

for `test.sql`:
```sql
DROP VIEW IF EXISTS test_view;
DROP TABLE IF EXISTS test_table_1;
DROP TABLE IF EXISTS test_table_2;


CREATE TABLE IF NOT EXISTS test_table_1 (
    id INT AUTO_INCREMENT PRIMARY KEY,
    table_1_data VARCHAR(100) NOT NULL UNIQUE
)
COMMENT 'Some comment for first table.';

CREATE TABLE IF NOT EXISTS test_table_2
(
    id INT AUTO_INCREMENT PRIMARY KEY,
    some_data VARCHAR(100) NOT NULL,
    another_data VARCHAR(100) NOT NULL,
    UNIQUE (some_data, another_data)
)
COMMENT 'Tuple of some_data and another_data for some particular reason. This is rather long comment.';

/*
 *  Gathers all available data for all attributes.
 */
CREATE VIEW test_view
AS
    SELECT
        TT1.table_1_data, TT2.some_data, TT2.another_data
    FROM test_table_1 TT1
    INNER JOIN test_table_2 TT2 ON TT1.id = TT2.id;
```

which is otherwise rather proper mysql file.

**Comment by barrywhart** [comment]
I'll take an initial look at this to see what are the gaps. A couple things I see right away:
* Our `CREATE TABLE` syntax doesn't support `COMMENT`.
* I don't believe `CREATE VIEW` is supported at all.

**Comment by barrywhart** [comment]
I looked through the `test.sql` in detail. These are the gaps I see:
* Parsing column types that aren't simply a single word, e.g. `VARCHAR(100)`
* Add support for MySQL `AUTO_INCREMENT` column property
* Add support for MySQL `COMMENT` table and column properties. Reference: https://dev.mysql.com/doc/refman/8.0/en/create-table.html
* Add support for `CREATE VIEW`

I may be able to look at these next week. @alanmcruickshank: Would we want to put this stuff in the ANSI dialect?

**Comment by barrywhart** [comment]
I opened a [PR](https://github.com/alanmcruickshank/sqlfluff/pull/154) which addresses all of the above except `CREATE VIEW`. I'll look at that next.

**Comment by barrywhart** [comment]
Ok, with #154, the `test.sql` above _almost_ parses. There is one parse error which relates to the `JOIN` part of the view. I had to add parentheses around the `JOIN` condition. This is a separate issue, so I didn't want to address it in this PR, because there may already be a separate issue logged for this. Any thoughts, @alanmcruickshank?

So this didn't work:
```
CREATE VIEW test_view
AS
    SELECT
        TT1.table_1_data, TT2.some_data, TT2.another_data
    FROM test_table_1 TT1
    INNER JOIN test_table_2 TT2 ON TT1.id = TT2.id;
```

But this did:
```
CREATE VIEW test_view
AS
    SELECT
        TT1.table_1_data, TT2.some_data, TT2.another_data
    FROM test_table_1 TT1
    INNER JOIN test_table_2 TT2 ON (TT1.id = TT2.id);
```

**Comment by barrywhart** [comment]
Can this issue be closed now?

**Comment by alanmcruickshank** [comment]
Yep I think so! Well done.

**Comment by barrywhart** [comment]
Thanks, @alanmcruickshank. If there are other issues you'd like me to take on or help with, please let me know.
